<?xml version="1.0"?>
<launch>
  <arg name="arm_1_joint_cmd_pos_name" default="/lwa4p/arm_1_joint/cmd_pos" />
  <arg name="arm_2_joint_cmd_pos_name" default="/lwa4p/arm_2_joint/cmd_pos" />
  <arg name="arm_3_joint_cmd_pos_name" default="/lwa4p/arm_3_joint/cmd_pos" />
  <arg name="up_down_lower_limit" default="[-2.8]" />
  <arg name="up_down_upper_limit" default="[2.8]" />
  <arg name="left_right_lower_limit" default="[-2.8]" />
  <arg name="left_right_upper_limit" default="[2.8]" />
  <arg name="near_far_lower_limit" default="[-2.8, -2.8]" />
  <arg name="near_far_upper_limit" default="[2.8, 2.8]" />
  <arg name="joint_states_topic" default="/arm/joint_states" />
  <arg name="shoulder_frame" default="arm_base_link" />
  <arg name="tcp_frame" default="arm_tcp_link" />
  <arg name="joint_names" default="['arm_1_joint', 'arm_2_joint', 'arm_3_joint', 'arm_4_joint', 'arm_5_joint', 'arm_6_joint']" />
  <arg name="arm_3_joint_index" default="3" />
  <arg name="follow_joint_trajectory_param" default="/arm/arm_eff_traj_controller/follow_joint_trajectory" />
  <arg name="next_pos_delta_factor" default="0.7" />
  <arg name="next_pos_error_factor" default="0.33" />
  <arg name="should_use_max_pos_delta" default="False" />
  <arg name="max_pos_delta" default="0.35" />
  <arg name="map_phi_to_2pi" default="True" />
  <arg name="pred_pos_frame" default="arm_base_link" />
  <arg name="pred_pos_topic" default="/pred_pos_2" />
  <arg name="target_position_topic" default="/lwa4p/target_position" />
  <arg name="error_topic" default="/lwa4p/error" />
  <arg name="sphere_marker_pub_topic" default="/lwa4p/sphere_marker" />
  <arg name="base_network_class_data_pub_topic" default="/lwa4p/base_network_class_data_pub_topic" />
  <arg name="error_class_data_topic" default="/lwa4p/error_class_data_topic" />
  <arg name="feedback_data_pub_topic" default="/lwa4p/feedback_data_pub_topic" />
  <arg name="gazebo_target_model_name" default="unit_sphere_2" />
  <arg name="gazebo_target_base_x" default="1.0" />
<!--   <arg name="gazebo_target_base_y" default="-3.6" /> -->
  <arg name="gazebo_target_base_y" default="-0.6" />
  <arg name="gazebo_target_base_z" default="0.6" />
  <arg name="error_threshold" default="0.07" />

  <node name="pub_lwa4p_target_as_marker" pkg="target_reaching_nengo" type="pub_target_pos_as_vis_marker.py" output="screen">
    <param name="target_position_topic" value="$(arg target_position_topic)" />
    <param name="sphere_marker_pub_topic" value="$(arg sphere_marker_pub_topic)" />
  </node>

  <node name="pub_lwa4p_target_as_gazebo_sphere" pkg="target_reaching_nengo" type="pub_target_pos_as_gazebo_sphere.py" output="screen">
    <param name="target_position_topic" value="$(arg target_position_topic)" />
    <param name="gazebo_target_model_name" value="$(arg gazebo_target_model_name)" />
    <param name="gazebo_target_base_x" value="$(arg gazebo_target_base_x)" />
    <param name="gazebo_target_base_y" value="$(arg gazebo_target_base_y)" />
    <param name="gazebo_target_base_z" value="$(arg gazebo_target_base_z)" />
  </node>

  <node name="pub_lwa4p_pred_pos_as_target" pkg="target_reaching_nengo" type="pub_lwa4p_target_position.py" output="screen">
    <param name="pred_pos_frame" value="$(arg pred_pos_frame)" />
    <param name="pred_pos_topic" value="$(arg pred_pos_topic)" />
    <param name="target_position_topic" value="$(arg target_position_topic)" />
    <param name="error_topic" value="$(arg error_topic)" />
  </node>

  <node name="Main_TR_CL_nengo_node_lwa4p" pkg="target_reaching_nengo" type="main_TR_CL_node.py" output="screen">
    <param name="arm_1_joint_cmd_pos_name" value="$(arg arm_1_joint_cmd_pos_name)" />
    <param name="arm_2_joint_cmd_pos_name" value="$(arg arm_2_joint_cmd_pos_name)" />
    <param name="arm_3_joint_cmd_pos_name" value="$(arg arm_3_joint_cmd_pos_name)" />
    <rosparam param="up_down_lower_limit" subst_value="True">$(arg up_down_lower_limit)</rosparam>
    <rosparam param="up_down_upper_limit" subst_value="True">$(arg up_down_upper_limit)</rosparam>
    <rosparam param="left_right_lower_limit" subst_value="True">$(arg left_right_lower_limit)</rosparam>
    <rosparam param="left_right_upper_limit" subst_value="True">$(arg left_right_upper_limit)</rosparam>
    <rosparam param="near_far_lower_limit" subst_value="True">$(arg near_far_lower_limit)</rosparam>
    <rosparam param="near_far_upper_limit" subst_value="True">$(arg near_far_upper_limit)</rosparam>
    <param name="joint_states_topic" value="$(arg joint_states_topic)" />
    <param name="shoulder_frame" value="$(arg shoulder_frame)" />
    <param name="tcp_frame" value="$(arg tcp_frame)" />
    <rosparam param="joint_names" subst_value="True">$(arg joint_names)</rosparam>
    <param name="arm_3_joint_index" value="$(arg arm_3_joint_index)" />
    <param name="next_pos_delta_factor" value="$(arg next_pos_delta_factor)" />
    <param name="map_phi_to_2pi" value="$(arg map_phi_to_2pi)" />
    <param name="base_network_class_data_pub_topic" value="$(arg base_network_class_data_pub_topic)" />
    <param name="error_class_data_topic" value="$(arg error_class_data_topic)" />
    <param name="feedback_data_pub_topic" value="$(arg feedback_data_pub_topic)" />
    <param name="target_position_topic" value="$(arg target_position_topic)" />
    <param name="error_topic" value="$(arg error_topic)" />
    <param name="error_threshold" value="$(arg error_threshold)" />
    <param name="should_use_max_pos_delta" value="$(arg should_use_max_pos_delta)" />
    <param name="max_pos_delta" value="$(arg max_pos_delta)" />
    <param name="next_pos_error_factor" value="$(arg next_pos_error_factor)" />
  </node>

  <node name="lwa4p_mapping" pkg="target_reaching_nengo" type="target_reaching_to_lwa4p_mapping.py" output="screen">
    <param name="arm_1_joint_cmd_pos_name" value="$(arg arm_1_joint_cmd_pos_name)" />
    <param name="arm_2_joint_cmd_pos_name" value="$(arg arm_2_joint_cmd_pos_name)" />
    <param name="arm_3_joint_cmd_pos_name" value="$(arg arm_3_joint_cmd_pos_name)" />
    <param name="joint_states_topic" value="$(arg joint_states_topic)" />
    <rosparam param="joint_names" subst_value="True">$(arg joint_names)</rosparam>
    <param name="arm_3_joint_index" value="$(arg arm_3_joint_index)" />
    <param name="follow_joint_trajectory_param" value="$(arg follow_joint_trajectory_param)" />
  </node>
</launch>
